### PR TITLE
Client library for talking to Docker to be used by the worker

### PR DESCRIPTION
**Pull request by klopyrev**
@percyliang

### CODE REVIEW

File: worker/docker_client.py [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+                os.path.join(cert_path, 'key.pem'))
+            self._ssl_context.check_hostname = False
+
+        # Test to make sure that a connection can be established.
+        try:
+            self.test()
+        except DockerException:
+            print >> sys.stderr, """
+On Linux, a valid Docker installation should create a Unix socket at
+//var/run/docker.sock.
```
Comment (percyliang) [inline worker/docker_client.py]: Extra /?

File: worker/docker_client.py [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+            if create_image_response.status != 200:
+                raise DockerException(create_image_response.read())
+
+            # Wait for the download to finish.
+            while True:
+                loop_callback()
+                response = None
+                line = ''
+                while True:
+                    ch = create_image_response.read(1)
```
Comment (percyliang) [inline worker/docker_client.py:130]: What's going on here?  This looks kind of weird/hacky - you have to keep on parsing something as JSON?  You can't just read the entire line and parse it as JSON?  Please document.

File: worker/docker_client.py [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+            'BASHRC=$(pwd)/.bashrc',
+            'cd %s' % docker_bundle_path,
+            # Run as the user that owns the bundle directory. That way
+            # any new files are created as that user and not root.
+            'U_ID=$(stat -c %%u %s)' % docker_bundle_path,
+            'G_ID=$(stat -c %%g %s)' % docker_bundle_path,
+            'sudo -u \\#$U_ID -g \\#$G_ID -n bash -c ' +
+            '"[ -e $BASHRC ] && . $BASHRC; "' +
+            '"export HOME=%s; "' % docker_bundle_path +
+            '\'(%s) >stdout 2>stderr\'' % command.replace('\'', '\'"\'"\''),
```
Comment (percyliang) [inline worker/docker_client.py:167]: Does this work for any command with any sort of existing quotes?  This always makes me nervous...

File: worker/docker_client.py [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+                    break
+                if 'error' in response:
+                    raise DockerException(response['error'])
+
+    @wrap_exception('Unable to start Docker container')
+    def start_container(self, bundle_path, uuid, command, docker_image,
+                        request_network, dependencies):
+        # Set up the command.
+        docker_bundle_path = '/' + uuid
+        docker_command = [
```
Comment (percyliang) [inline worker/docker_client.py]: rename to 'docker_commands', since it's an array

**Comment by percyliang** [comment]
Does this code only work with certain versions of Docker?  If so, we should check the docker version and throw an error if we don't have the right version.

File: worker/docker_client.py (hunk 1 above)
Comment (klopyrev) [inline worker/docker_client.py]: Done.

File: worker/docker_client.py (hunk 2 above)
Comment (klopyrev) [inline worker/docker_client.py:130]: Added more documentation. It's just one way of parsing things.

File: worker/docker_client.py (hunk 3 above)
Comment (klopyrev) [inline worker/docker_client.py:167]: Added more documentation. I tested this fairly well. For example, including things like:
-  Running multiple commands a time
- Running echo $HOME -> returns  /0x??????????????
- Running echo '$HOME' -> returns $HOME

File: worker/docker_client.py (hunk 4 above)
Comment (klopyrev) [inline worker/docker_client.py]: Done.

**Comment by percyliang** [comment]
LGTM

**Comment by klopyrev** [comment]
Done.

**Comment by klopyrev** [comment]
Also, I added another commit that fixes two small bugs.

**Pull request merged**